File: platforms/qt/apps/BrisaMediaRenderer/avtransport.cpp
#include "avtransport.h"

using namespace Brisa;
using namespace Phonon;

AVTransport::AVTransport(Phonon::MediaObject *player) :
        BrisaService("urn:schemas-upnp-org:service:AVTransport:1",
                     "AVTransport",
                     "/AVTransport/render-transport-scpd.xml",
                     "/AVTransport/control",
                     "/AVTransport/eventSub")
{
    this->player = player;
    setDescriptionFile("render-transport-scpd.xml");
    this->manager = new QNetworkAccessManager();
}

AVTransport::~AVTransport()
{
    delete this->player;
    delete this->manager;
}

BrisaOutArgument* AVTransport::setavtransporturi(
        BrisaInArgument *const inArguments, BrisaAction *const action)
{
    Q_UNUSED(action)

    if (!inArguments->contains("CurrentURI"))
        return new BrisaOutArgument();

    connect(manager, SIGNAL(finished(QNetworkReply*)), this, SLOT(downloadFinished(QNetworkReply*)));

    QUrl url(inArguments->value("CurrentURI"));
    manager->get(QNetworkRequest(url));
    getVariable("AVTransportURI")->setAttribute(BrisaStateVariable::Value,
                                                     inArguments->value("CurrentURI"));
    getVariable("TransportState")->setAttribute(BrisaStateVariable::Value, "STOPPED");
    getVariable("NumberOfTracks")->setAttribute(BrisaStateVariable::Value, "1");

    return new BrisaOutArgument();
}

BrisaOutArgument* AVTransport::getmediainfo(BrisaInArgument *const inArguments,
                                            BrisaAction *const action)
{
    Q_UNUSED(inArguments)
    Q_UNUSED(action)
    BrisaOutArgument *outArgs = new BrisaOutArgument();
    outArgs->insert("NrTracks",
                   getVariable("NumberOfTracks")->getValue().toString());

    outArgs->insert("MediaDuraction",
                   getVariable("CurrentMediaDuration")->getValue().toString());

    outArgs->insert("CurrentURI",
                   getVariable("AVTransportURI")->getValue().toString());

    outArgs->insert("NextURI",
                   getVariable("NextAVTransportURI")->getValue().toString());

    outArgs->insert("PlayMedium",
                   getVariable("PlaybackStorageMedium")->getValue().toString());

    outArgs->insert("CurrentURIMetaData", "");
    outArgs->insert("NextURIMetaData", "");
    outArgs->insert("RecordMedium", "");
    outArgs->insert("WriteStatus", "");
    return outArgs;
}

BrisaOutArgument* AVTransport::gettransportinfo(BrisaInArgument *const inArguments,
                                                BrisaAction *const action)
{
    Q_UNUSED(inArguments)
    Q_UNUSED(action)
    //TODO: everything!
    //InstanceID is an input argument.
    BrisaOutArgument *outArgs = new BrisaOutArgument();

    outArgs->insert("NrTracks",
                   getVariable("NumberOfTracks")->getValue().toString());
    outArgs->insert("CurrentTransportState",
                   getVariable("TransportState")->getValue().toString());
    outArgs->insert("CurrentTransportStatus", "");
    outArgs->insert("CurrentSpeed", "");
    return outArgs;
}

BrisaOutArgument* AVTransport::getpositioninfo(
        BrisaInArgument *const inArguments, BrisaAction *const action)
{
    Q_UNUSED(inArguments)
    Q_UNUSED(action)
    //TODO: everything!
    //InstanceID is an input argument.
    BrisaOutArgument *outArgs = new BrisaOutArgument();
    outArgs->insert("Track", "");
    outArgs->insert("TrackDuration", "");
    outArgs->insert("TrackMetaData", "");
    outArgs->insert("TrackURI", "");
    outArgs->insert("RelTime", "");
    outArgs->insert("AbsTime", "");
    outArgs->insert("RelCount", "");
    outArgs->insert("AbsCount", "");
    return outArgs;
}

BrisaOutArgument* AVTransport::getdevicecapabilities(
        BrisaInArgument *const inArguments, BrisaAction *const action)
{
    Q_UNUSED(inArguments)
    Q_UNUSED(action)
    //TODO: everything!
    //InstanceID is an input argument.
    BrisaOutArgument *outArgs = new BrisaOutArgument();
    outArgs->insert("PlayMedia", "");
    outArgs->insert("RecMedia", "");
    outArgs->insert("RecQualityModes", "");
    return outArgs;
}

BrisaOutArgument* AVTransport::gettransportsettings(
        BrisaInArgument *const inArguments, BrisaAction *const action)
{
    Q_UNUSED(inArguments)
    Q_UNUSED(action)
    //TODO: everything!
    //InstanceID is an input argument.
    BrisaOutArgument *outArgs = new BrisaOutArgument();
    outArgs->insert("PlayMode", "");
    outArgs->insert("RecQualityMode", "");
    return outArgs;
}

BrisaOutArgument* AVTransport::stop(BrisaInArgument *const inArguments,
                                    BrisaAction *const action)
{
    Q_UNUSED(inArguments)
    Q_UNUSED(action)
    //TODO: everything!
    //InstanceID is an input argument.
    //Speed is an input argument.

    this->player->stop();
    getVariable("TransportState")->setAttribute(BrisaStateVariable::Value, "STOPPED");

    return new BrisaOutArgument();
}

BrisaOutArgument* AVTransport::play(BrisaInArgument *const inArguments,
                                    BrisaAction *const action)
{
    Q_UNUSED(inArguments)
    Q_UNUSED(action)
    //InstanceID is an input argument.
    //Speed is an input argument.

    QString uri = getVariable("AVTransportURI")->getAttribute(BrisaStateVariable::Value);

    qDebug() << this->getVariable("AVTransportURI")->getAttribute(BrisaStateVariable::Value);
//    this->player->setCurrentSource(MediaSource(QUrl(
//            this->getVariable("AVTransportURI")->getAttribute(
//                    BrisaStateVariable::Value))));
//    this->player->play();
    if (getVariable("TransportState")->getAttribute(BrisaStateVariable::Value) == "STOPPED") {
        if (uri.contains("http")) {
            player->setCurrentSource(MediaSource(QUrl(uri)));
        } else {
            player->setCurrentSource(MediaSource(uri));
        }
    }

    player->play();
    getVariable("TransportState")->setAttribute(BrisaStateVariable::Value, "PLAYING");

    return new BrisaOutArgument();
}

BrisaOutArgument* AVTransport::pause(BrisaInArgument *const inArguments,
                                     BrisaAction *const action)
{
    Q_UNUSED(inArguments)
    Q_UNUSED(action)
    //InstanceID is an input argument.

    player->pause();
    getVariable("TransportState")->setAttribute(BrisaStateVariable::Value,
                                                     "PAUSED_PLAYBACK");
    return new BrisaOutArgument();
}

BrisaOutArgument* AVTransport::seek(BrisaInArgument *const inArguments,
                                    BrisaAction *const action)
{
    Q_UNUSED(action)
    //InstanceID is an input argument.
    //Unit is an input argument.
    //Target is an input argument.

    if (inArguments->value("Unit") == "ABS_TIME")
        player->seek(inArguments->value("Unit").toInt());

    return new BrisaOutArgument();
}

BrisaOutArgument* AVTransport::next(BrisaInArgument *const inArguments,
                                    BrisaAction *const action)
{
    Q_UNUSED(inArguments)
    Q_UNUSED(action)
    //TODO: Everything!
    //InstanceID is an input argument.

    return new BrisaOutArgument();
}

BrisaOutArgument* AVTransport::previous(BrisaInArgument *const inArguments,
                                        BrisaAction *const action)
{
    Q_UNUSED(inArguments)
    Q_UNUSED(action)
    //TODO: Everything!
    //InstanceID is an input argument.

    return new BrisaOutArgument();
}

void AVTransport::downloadFinished(QNetworkReply *reply)
{
    QByteArray data = reply->readAll().data();
    this->currentURI = data;
    qDebug() << "Downloaded: " << data;
}

File: platforms/python/tags/python-brisa-0.3.0/src/media_services/avtransport/av_transport.py
# Licensed under the MIT license
# http://opensource.org/licenses/mit-license.php
# Copyright 2007 Brisa Team <brisa-develop@garage.maemo.org>

#TODO: Implement the required Actions

#REQUIRED ACTIONS:
#SetAVTransportURI()     R
#GetMediaInfo()          R
#GetMediaInfo_Ext()      R
#GetTransportInfo()      R
#GetPositionInfo()       R
#GetDeviceCapabilities() R
#GetTransportSettings()  R
#Stop()                  R
#Play()                  R
#Seek()                  R
#Next()                  R
#Previous()              R

#TODO: Handle errors (transport_staus)

from twisted.python import log

from upnp import ServiceControl
from upnp import errorCode
from media_services.gst_renderer import GSTRendererMaemo, GSTRenderer
import platform

class AVTransportControl(ServiceControl):


    def __init__(self):
        self.avTransportURI = None
        self.NumberOfTracks = 0

        if platform.uname()[4].find('arm') is not -1:
            self.gst_player = GSTRendererMaemo()
        elif platform.system() == 'Linux':
            self.gst_player = GSTRenderer()

        self.urilist = {}
        self.transport_state = 'NO_MEDIA_PRESENT'
        self.transport_status= 'OK'
        self.transport_speed = 1

    def soap_SetAVTransportURI(self, *args, **kwargs):
        """This action specifies the URI of the resource to be controlled by the specified
        AVTransport instance. It is RECOMMENDED that the AVTransport service checks the MIME-type
        of the specified resource when executing this action"""

        (InstanceID, CurrentURI, CurrentURIMetaData) = args

        self.avTransportURI = CurrentURI
        self.NumberOfTracks = 1
        self.gst_player.av_uri = CurrentURI
        self.transport_state = 'STOPPED'

        log.msg('SetAVTransportURI()')

    def soap_GetMediaInfo(self, *args, **kwargs):
        """This action returns information associated with the current media of the
        specified instance; it has no effect on state."""

        log.msg('GetMediaInfo()')

        return {'GetMediaInfoResponse': {'NrTracks': str(self.NumberOfTracks), 'MediaDuration': str(self.gst_player.queryDuration()[0]),
                                                     'CurrentURI': str(self.avTransportURI),
                                                     'CurrentURIMetaData': '',
                                                     'NextURI':'', 'NextURIMetaData':'',
                                                     'PlayMedium':'NETWORK',
                                                     'RecordMedium':'',
                                                     'WriteStatus': ''}}


    def soap_GetMediaInfo_Ext(self, *args, **kwargs):
        """This action returns information associated with the current media of the specified
        instance; it has no effect on state.The information returned is identical to the information
        returned by the GetMediaInfo() action, except for the additionally returned CurrentType
        argument """

        log.msg('GetMediaInfo_Ext()')

        return {'GetMediaInfo_ExtResponse': {'CurrentType' : 'TRACK_UNAWARE',
                                                     'NrTracks': str(self.NumberOfTracks),
                                                     'MediaDuration': str(self.gst_player.queryDuration()[0]),
                                                     'CurrentURI': str(self.avTransportURI),
                                                     'CurrentURIMetaData': '',
                                                     'NextURI':'', 'NextURIMetaData':'',
                                                     'PlayMedium':'NETWORK',
                                                     'RecordMedium':'',
                                                     'WriteStatus': ''}}


    def soap_GetTransportInfo(self, *args, **kwargs):
        """This action returns information associated with the current transport state
        of the specified instance; it has no effect on state."""

        #log.msg('GetTransportInfo()')

        self.transport_state = self.gst_player.getState()

        #print self.transport_state
        #print self.transport_status

        return {'GetTransportInfoResponse': {'CurrentTransportState': str(self.transport_state),
                                                     'CurrentTransportStatus': str(self.transport_status),
                                                     'CurrentSpeed': str(self.transport_speed)}}


    def soap_GetPositionInfo(self, *args, **kwargs):
        """This action returns information associated with the current position of
        the transport of the specified instance; it has no effect on state."""

        #log.msg('GetPositionInfo()')

        dur = self.gst_player.queryDuration()
        pos = self.gst_player.queryPosition()

        duration = dur[0]

        if dur[1] == -1:
            duration = pos[0]


        abs_pos = pos[1]
        rel_pos = pos[1]

        if dur[1] > 2147483646:
            abs_pos = pos[1]*2147483646/dur[1]

        if pos[1] == -1:
            abs_pos = 0
            rel_pos = 0

        return {'GetPositionInfoResponse':{'Track':'1',
                                           'TrackDuration':'"'+str(duration)+'"',
                                           'TrackMetaData':'',
                                           'TrackURI':str(self.avTransportURI),
                                           'RelTime':'"'+str(pos[0])+'"',
                                           'AbsTime':'"'+str(pos[0])+'"',
                                           'RelCount':rel_pos,
                                           'AbsCount':abs_pos}}

    def soap_GetDeviceCapabilities(self, *args, **kwargs):
        """This action returns information on device capabilities of the specified instance,
        such as the supported playback and recording formats, and the supported quality levels for
        recording. This action has no effect on state."""

        log.msg('GetDeviceCapabilities()')

    def soap_GetTransportSettings(self, *args, **kwargs):
        """This action returns information on various settings of the specified instance, such as
        the current play mode and the current recording quality mode. This action has no
        effect on state."""

        log.msg('GetDeviceCapabilities()')

    def soap_Stop(self, *args, **kwargs):
        """This action stops the progression of the current resource that is associated with the
        specified instance."""

        log.msg('Stop()')
        self.transport_state = 'TRANSITIONING'
        self.gst_player.stop()
        self.transport_state = 'STOPPED'

    def soap_Play(self, *args, **kwargs):
        """This action starts playing the resource of the specified instance, at the specified speed,
        starting at the current position, according to the current play mode."""

        log.msg('Play()')

        self.transport_state = 'TRANSITIONING'

        self.gst_player.play()

        self.transport_state = 'PLAYING'

    def soap_Pause(self, *args, **kwargs):
        """This action starts playing the resource of the specified instance, at the specified speed,
        starting at the current position, according to the current play mode."""

        log.msg('Pause()')

        self.transport_state = 'TRANSITIONING'

        self.gst_player.pause()

        self.transport_state = 'PAUSED_PLAYBACK'

    def soap_Seek(self, *args, **kwargs):
        """This action starts seeking through the resource controlled by the specified instance - as
        fast as possible - to the position, specified in the Target argument."""

        (InstanceID, Unit, Target) = args

        self.gst_player.seek(Unit, Target)

        log.msg('Seek()')

    def soap_Next(self, *args, **kwargs):
        """This is a convenient action to advance to the next track. This action is functionally equivalent
        to Seek(TRACK_NR, CurrentTrackNr+1). This action does not cycle back to the first track."""

        log.msg('Next()')

    def soap_Previous(self, *args, **kwargs):
        """This is a convenient action to advance to the previous track. This action is functi
